Prevent the single-node tree check from yielding "false positives" on linear trees; also fix the build

fedarko opened the pull request:
## Background

The check for single-node trees was previously structured as follows (where `data` is a Newick-format string representing a tree):

```python
    if data.count(',') == 0:
        raise ValueError("Only trees with more than 1 node supported")
```

In most cases, this check works. However, a Newick file containing no commas does not indicate that this tree contains only one node -- it just indicates that this tree does not contain any branches. In the weird case where our "tree" is a linear chain of nodes without any branches, e.g. `((b:3)a:2)root:1;`, this check will falsely raise an error. (I don't expect that trees like this will come up in practice often, but some of the trees in Empress' tests are structured like this -- so the current check is breaking Empress' build.)

## Changes in this PR

This PR fixes this check: now it is based on the topology of the parsed tree, rather than the input Newick string's number of commas. This means that the check will be performed "later on", from the user's perspective, but single-node trees should by definition be parsed pretty quickly anyway -- so I think postponing the check is reasonable.

This PR also adds a test, `test_parse_newick_linear_tree`, which ensures that "linear trees" like the ones discussed here can be parsed ok.

Also, the Travis build failed for this test (see log [here](https://app.travis-ci.com/github/fedarko/improved-octo-waddle/builds/249606503), which doesn't seem to show up on this PR but I did get an email about?) -- it seems like this failure is due to 1) `BitArray` being moved to within the `bp` folder, which broke `make test`, and 2) flake8 issues. This PR additionally resolves both of these concerns, so the Travis build now [works again](https://app.travis-ci.com/github/fedarko/improved-octo-waddle/builds/249607324).

Thanks!

fedarko committed:
BUG: Fix single-node-tree check

Reliant now on the topology rather than on the string, since relying
on the presence of "," yields false positives in the case of a tree
with multiple nodes in a single linear chain.

fedarko committed:
TST: fix tests (new BitArray location)

fedarko committed:
STY/TST: fix flake8 issues; add linear tree test

fedarko committed:
TST: Explicitly verify that the only non-tip is a

Just a small check that linear trees somehow aren't parsed as having
*more* nodes than they should. IDK how that would even happen, but
this ensures it shouldn't at least...

wasade commented:
Looks good, thanks! I'm surprised actions isn't running -- I'll need to look into that before merge

wasade commented:
I'm confused about travis as the project was using gh actions as of a few weeks ago, and I haven't made any changes to its configuration since

fedarko committed:
Merge branch 'master' of https://github.com/biocore/improved-octo-waddle into single-node-falsepositive-fix

fedarko committed:
BLD: remove travis

To avoid confusion; the CI for iow is now on github actions

fedarko committed:
DOC/BLD: add the (obligatory) fancy badge

fedarko commented:
@wasade got it! I've merged master in, and now it looks like github actions is able to run. Looks like all of the builds are successful.

There was a `.travis.yml` file remaining in the repository, so -- at least for my fork of iow -- Travis was still trying to run builds, which is weird (yesterday I got emails from Travis about the builds of my fork of iow failing, and I assumed incorrectly that iow was just using Travis instead of github actions?). Anyway, I added another commit that removes `.travis.yml` to avoid confusion (and relieve the stress on Travis' servers, I guess :) I also added one of those fancy badges to the README that reflects the github actions CI status, to make things clearer for other devs.

ElDeveloper commented on Makefile:
```diff
@@ -1,6 +1,6 @@
 libbitarr.a: 
-	cd BitArray && $(MAKE)
-	make -C BitArray -f Makefile
+	cd bp/BitArray && $(MAKE)
+	make -C bp/BitArray -f Makefile
```
@fedarko, any chance you can explain why this needed to be added and why building the project worked before these changes? Maybe I'm missing something but nothing else changed, right?

fedarko replied:
> why building the project worked before these changes

I'm not 100% sure on how the build process for iow works. My understanding is that building the project succeeded because it didn't just call `make build` or `make libbitarr.a`: instead, the setup script calls

https://github.com/biocore/improved-octo-waddle/blob/2f92d78d5f1f68168c3ba1adf6979603a257a592/setup.py#L41-L46

The `make -C bitarr` thing tells `make` to move to the directory referred to by `bitarr`(`bp/BitArray`) before running this command. Although I feel like this should fail for another reason, because then `make libbitarr.a` tries to cd into `BitArray` again, and this directory does not exist (we're already in `BitArray`, and [there shouldn't be another `BitArray` directory underneath it](https://github.com/noporpoise/BitArray/tree/f444eb45ab3cac67d00089c50f6bb42729838db8)). So... I got nothing?

> why this needed to be added

The main reason I added this fix in this PR was not because of the build but because it broke `make test`. I looked into it, because the GitHub Actions build was passing before -- the reason it was passing before this fix is because **`make test` is not actually how GH Actions runs its tests**. Instead, GH Actions just uses `pytest`:

https://github.com/biocore/improved-octo-waddle/blob/2f92d78d5f1f68168c3ba1adf6979603a257a592/.github/workflows/python-package-conda.yml#L40-L43

This is separate from how `make test` works:

https://github.com/biocore/improved-octo-waddle/blob/2f92d78d5f1f68168c3ba1adf6979603a257a592/Makefile#L1-L11

It'd probably be ideal to combine these into a single test-running command, but I'm not sure how to do that (and that is likely out of scope for this PR).

----

### TLDR

If you'd like, I can remove this change for now -- just to avoid increasing weird issues with installation. As a separate issue, I think it'd make sense to change `make test` to reflect the GitHub Actions stuff (i.e. just running `pytest`)?

ElDeveloper reviewed (changes requested):
Looks good to me, just one comment.

wasade commented on bp/_io.pyx:
```diff
@@ -148,16 +148,16 @@ cpdef parse_newick(unicode data):
     datalen = len(data)
     topology = _newick_to_bp(data)
 
+    if len(topology.B) <= 1:
```
This is not correct, a single node would be `[1, 0]` in the structure. Adjusted on the subsequent PR which includes the relevant changes here

wasade commented:
Closing in favor of #61, sorry for the delay.
